commit 4f7637304eca894adf2e70078a55654a88224a30
Author: Lukasz Anforowicz <lukasza@chromium.org>
Date:   Tue Sep 30 07:08:03 2025 -0700

    [rust png] Reland: Reinforce `testonly`-ness of some users of `libpng`.
    
    This reverts commit 7fa4c2d7ab42a48247ce8e0290cbb7e854990f1e and
    effectively relands https://crrev.com/c/6976870.  The reland is
    attempted after using wildcards to cover extra gni-generated targets.
    `Cq-Include-Trybots` directives in the CL description should verify
    correctness of the reland.  Original CL description follows:
    
    This CL extracts a separate `third_party/libpng:libpng_for_testonly`
    target which:
    
    1) Is a `group` that proxies/forwards to the `libpng` target by listing
       it as its `public_deps`
    2) Has limited visibility (based on the new `libpng_testonly_visibility`
       introduced in the refactored `third_party/libpng/visibility.gni`).
       The new visibility is almost unchanged, except for explicitly
       spelling out some targets instead of using wildcards
    3) Is marked as `testonly`
    
    Bug: 443128323
    Change-Id: If7ba0d184324a5c662aaf2d52122994ea778d452
    Cq-Include-Trybots: luci.chromium.try:ios-catalyst
    Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6991491
    Auto-Submit: Łukasz Anforowicz <lukasza@chromium.org>
    Commit-Queue: Łukasz Anforowicz <lukasza@chromium.org>
    Reviewed-by: Nico Weber <thakis@chromium.org>
    Cr-Commit-Position: refs/heads/main@{#1522830}

diff --git a/chrome/test/BUILD.gn b/chrome/test/BUILD.gn
index a4703c3692d..2d70f7a5fe9 100644
--- a/chrome/test/BUILD.gn
+++ b/chrome/test/BUILD.gn
@@ -11152,7 +11152,7 @@ if (!is_android) {
         "//testing/gtest",
         "//third_party/hunspell",
         "//third_party/icu",
-        "//third_party/libpng",
+        "//third_party/libpng:libpng_for_testonly",
         "//ui/base:test_support",
         "//ui/events:test_support",
         "//ui/ozone",
@@ -11512,7 +11512,7 @@ if (!is_android && !is_chromeos_device) {
       "//testing/gtest",
       "//third_party/hunspell",
       "//third_party/icu",
-      "//third_party/libpng",
+      "//third_party/libpng:libpng_for_testonly",
       "//ui/base:base_interactive_ui_tests",
       "//ui/base:ozone_buildflags",
       "//ui/base:test_support",
@@ -12470,7 +12470,7 @@ if (!is_android) {
       "//testing/gtest",
       "//third_party/hunspell",
       "//third_party/icu",
-      "//third_party/libpng",
+      "//third_party/libpng:libpng_for_testonly",
       "//ui/base:test_support",
       "//ui/resources:ui_test_pak",
       "//ui/views",
diff --git a/testing/libfuzzer/fuzzers/BUILD.gn b/testing/libfuzzer/fuzzers/BUILD.gn
index 7d242e177f0..1c67eee1598 100644
--- a/testing/libfuzzer/fuzzers/BUILD.gn
+++ b/testing/libfuzzer/fuzzers/BUILD.gn
@@ -104,7 +104,7 @@ fuzzer_test("libpng_progressive_read_fuzzer") {
   sources = [ "libpng_read_fuzzer.cc" ]
   deps = [
     "//base",
-    "//third_party/libpng",
+    "//third_party/libpng:libpng_for_testonly",
   ]
   dict = "dicts/png.dict"
   seed_corpuses = libpng_seed_corpuses
diff --git a/third_party/libpng/BUILD.gn b/third_party/libpng/BUILD.gn
index d4bdea1e33c..3e31348b94f 100644
--- a/third_party/libpng/BUILD.gn
+++ b/third_party/libpng/BUILD.gn
@@ -130,6 +130,14 @@ if (is_win) {
   }
 }
 
+group("libpng_for_testonly") {
+  testonly = true
+  public_deps = [ ":libpng" ]
+
+  visibility = []
+  visibility = libpng_testonly_visibility
+}
+
 if (build_with_chromium) {
   libpng_ossfuzz_seed_corpuses = [
     "//components/test/data/viz",
diff --git a/tools/imagediff/BUILD.gn b/tools/imagediff/BUILD.gn
index d56651bedc7..2b5f3d38488 100644
--- a/tools/imagediff/BUILD.gn
+++ b/tools/imagediff/BUILD.gn
@@ -34,6 +34,7 @@ if (target_os == "win" && host_os != "win") {
 # If the current toolchain is the test host toolchain, build the tool.
 if (current_toolchain == imagediff_toolchain) {
   executable("imagediff") {
+    testonly = true
     output_name = "image_diff"  # Different than dir name for historical
                                 # reasons.
     sources = [
@@ -47,7 +48,7 @@ if (current_toolchain == imagediff_toolchain) {
     deps = [
       "//base",
       "//build/win:default_exe_manifest",
-      "//third_party/libpng",
+      "//third_party/libpng:libpng_for_testonly",
       "//third_party/zlib",
     ]
   }
@@ -60,6 +61,7 @@ if (current_toolchain == imagediff_toolchain) {
 } else if (current_toolchain == default_toolchain &&
            default_toolchain != imagediff_toolchain) {
   binary_symlink("imagediff") {
+    testonly = true
     binary_label = ":$target_name($imagediff_toolchain)"
     binary_output_name = "image_diff"
 
diff --git a/ui/gfx/BUILD.gn b/ui/gfx/BUILD.gn
index f6a40abfe99..83d8f55027a 100644
--- a/ui/gfx/BUILD.gn
+++ b/ui/gfx/BUILD.gn
@@ -906,7 +906,7 @@ test("gfx_unittests") {
     "//skia:skcms",
     "//testing/gtest",
     "//third_party/icu:icuuc",
-    "//third_party/libpng",
+    "//third_party/libpng:libpng_for_testonly",
     "//third_party/zlib",
     "//ui/base",
     "//ui/gfx/animation",
